src/linkediting.js

```diff
@@ -55,5 +60,90 @@ export default class LinkEditing extends Plugin {
 
 		// Enable two-step caret movement for `linkHref` attribute.
 		bindTwoStepCaretToAttribute( editor.editing.view, editor.model, this, 'linkHref' );
+
+		// Setup highlight over selected link.
+		this._setupLinkHighlight();
+	}
+
+	/**
+	 * Adds highlight over link which has selection inside, together with two-step caret movement indicates whenever
+	 * user is typing inside the link.
+	 *
+	 * @private
+	 */
+	_setupLinkHighlight() {
+		const editor = this.editor;
+		const model = this.editor.model;
+		const doc = model.document;
+		const highlightDescriptor = {
+			id: 'linkBoundaries',
+			class: 'ck-link_selected',
+			priority: 1
+		};
+
+		// Convert linkBoundaries marker to view highlight.
+		editor.conversion.for( 'editingDowncast' )
+			.add( downcastMarkerToHighlight( {
+				model: 'linkBoundaries',
+				view: highlightDescriptor
+			} ) );
+
+		// Create marker over whole link when selection has "linkHref" attribute.
+		doc.on( 'change', () => {
+			const selection = doc.selection;
+
+			// Create marker over link when selection is inside or remove marker otherwise.
+			if ( selection.hasAttribute( 'linkHref' ) ) {
+				const modelRange = findLinkRange( selection.getFirstPosition(), selection.getAttribute( 'linkHref' ) );
+				const marker = model.markers.get( 'linkBoundaries' );
+
+				if ( !marker || !marker.getRange().isEqual( modelRange ) ) {
+					model.change( writer => {
+						writer.setMarker( 'linkBoundaries', modelRange );
+					} );
+				}
+			} else if ( model.markers.has( 'linkBoundaries' ) ) {
+				model.change( writer => {
+					writer.removeMarker( 'linkBoundaries' );
+				} );
+			}
+		} );
+
+		// Custom converter for selection's "linkHref" attribute - when collapsed selection has this attribute it is
+		// wrapped with <span> similar to that used by highlighting mechanism. This <span> will be merged together with
+		// highlight wrapper. This prevents link splitting When selection is at the beginning or at the end of the link.
+		// Without this converter:
+		//
+		//		<a href="url">{}</a><span class="ck-link_selected"><a href="url">foo</a></span>
+		//
+		// When converter is applied:
+		//
+		//		<span class="ck-link_selected"><a href="url">{}foo</a></span>
+		editor.editing.downcastDispatcher.on( 'attribute:linkHref', ( evt, data, conversionApi ) => {
+			const selection = data.item;
+
+			if ( !( selection instanceof DocumentSelection || selection instanceof ModelSelection ) || !selection.isCollapsed ) {
+				return;
+			}
+
+			const writer = conversionApi.writer;
+			const viewSelection = writer.document.selection;
+			const wrapper = new HighlightAttributeElement( 'span', { class: 'ck-link_selected' } );
+			wrapper._priority = highlightDescriptor.priority;
+			wrapper._setCustomProperty( 'highlightDescriptorId', highlightDescriptor.id );
```

Comment:
Use `writer` instead of protected method.

Reply:
This whole code will be simplified after the changes in markers will land in engine. We can create a follow-up to clean this.